### ISSUE TITLE
Entered into a loop "Error: This socket has been ended by the other party"

### ISSUE THREAD

**Issue by ibc** [theturtle32/WebSocket-Node#294]
I'm not sure whether this is an issue in WebSocket-Node, or in my [library](https://github.com/ibc/protoo/) using it, or even in my application using my library. However I don't find anything in my code that could theoretically produce it.

Basically, after months using it without issues, I've got an infinite loop that blocked the Node process:

```
Error: This socket has been ended by the other party
Error: This socket has been ended by the other party
Error: This socket has been ended by the other party
Error: This socket has been ended by the other party
Error: This socket has been ended by the other party
Error: This socket has been ended by the other party
...
```

I don't run any kind of `setInterval` or similar when I try to close the WS transport or when I try to send a message over it, so I expect this is an strange race condition or corner case when some WS entity emits `close` event and its `close()` method is later called (or vice-versa). And such an entity may be the `WebSocketConnection` (or not).

It seems to me that this "corner case" happened after the client sent a TCP FIN so it was half-closed.

Basically this is the code within my app that deals with the WebSocket-Node `WebSocketConnection` object:

* [Peer](https://github.com/ibc/protoo/blob/master/server/lib/Peer.js#L108): Just a wrapper over my lib's WebSocketTransport.
* [WebSocketTransport](https://github.com/ibc/protoo/blob/master/server/lib/transports/WebSocketTransport.js#L62): A kind of wrapper over Node-WebSocket `WebSocketConnection`.

Something coming to your mind?

**Comment by ibc** [comment]
For example, I'm a bit afraid about the following code:

```js
WebSocketConnection.prototype._addSocketEventListeners = function() {
    this.socket.on('error', this.handleSocketError.bind(this));
    this.socket.on('end', this.handleSocketEnd.bind(this));
    this.socket.on('close', this.handleSocketClose.bind(this));
    this.socket.on('drain', this.handleSocketDrain.bind(this));
    this.socket.on('pause', this.handleSocketPause.bind(this));
    this.socket.on('resume', this.handleSocketResume.bind(this));
    this.socket.on('data', this.handleSocketData.bind(this));
};

WebSocketConnection.prototype.handleSocketError = function(error) {
    this._debug('handleSocketError: %j', error);
    this.closeReasonCode = WebSocketConnection.CLOSE_REASON_ABNORMAL;
    this.closeDescription = 'Socket Error: ' + error.syscall + ' ' + error.code;
    this.connected = false;
    this.state = STATE_CLOSED;
    this.fragmentationSize = 0;
    if (utils.eventEmitterListenerCount(this, 'error') > 0) {
        this.emit('error', error);
    }
    this.socket.destroy(error);
    this._debug.printOutput();
};

WebSocketConnection.prototype.handleSocketEnd = function() {
    this._debug('handleSocketEnd: received socket end.  state = %s', this.state);
    this.receivedEnd = true;
    if (this.state === STATE_CLOSED) {
        // When using the TLS module, sometimes the socket will emit 'end'
        // after it emits 'close'.  I don't think that's correct behavior,
        // but we should deal with it gracefully by ignoring it.
        this._debug('  --- Socket \'end\' after \'close\'');
        return;
    }
    if (this.state !== STATE_PEER_REQUESTED_CLOSE &&
        this.state !== STATE_ENDING) {
      this._debug('  --- UNEXPECTED socket end.');
      this.socket.end();
    }
};
```

I've tested that, even if already called before, a new call to `socket.destroy(error)` produces the `error` event again. Calling `socket.end()` multiple times also produces a `close` event in the socket.

What I mean is that, within the above code, I may perceive an open door to "close" / "error" / "end" / "destroy" race conditions, with events calling methods that produce them again, etc. However I've not able to manually reproduce it by calling them (but note that I haven't tried to half-close the client connection before...).

**Comment by ibc** [comment]
I've been able to reproduce this issue 10 times today. It seems that this will be a hard problem for me.

**Comment by ibc** [comment]
Definitely it seems to me that this issue is 100% related to #288, fixed in #289.

**Comment by Sean-Der** [comment]
@ibc I seem to still be hitting this

```
sean@SeanLaptop:node_modules$ npm list | grep webso
│ └─┬ websocket@1.0.25
│ └── websocket@1.0.25 deduped
```

```
Oct 19 10:52:10 webrtc0-stage node[1408]: Thu, 19 Oct 2017 10:52:10 GMT protoo-server:ERROR:WebSocketTransport connection "error" event [conn:WSS:[**.208.196.92]:55402, error:Error: This socket has been ended by the other party]
Oct 19 10:52:10 webrtc0-stage node[1408]: Thu, 19 Oct 2017 10:52:10 GMT protoo-server:ERROR:WebSocketTransport connection "error" event [conn:WSS:[**.208.196.92]:55402, error:Error: This socket has been ended by the other party]
Oct 19 10:52:10 webrtc0-stage node[1408]: Thu, 19 Oct 2017 10:52:10 GMT protoo-server:ERROR:WebSocketTransport connection "error" event [conn:WSS:[**.208.196.92]:55402, error:Error: This socket has been ended by the other party]
Oct 19 10:52:10 webrtc0-stage node[1408]: Thu, 19 Oct 2017 10:52:10 GMT protoo-server:ERROR:WebSocketTransport connection "error" event [conn:WSS:[**.208.196.92]:55402, error:Error: This socket has been ended by the other party]
Oct 19 10:53:26 webrtc0-stage node[1408]: <--- Last few GCs --->
Oct 19 10:53:26 webrtc0-stage node[1408]: [1408:0x33e1bc0]  1353382 ms: Mark-sweep 1405.3 (1462.9) -> 1405.2 (1436.9) MB, 899.3 / 0.0 ms  (+ 0.0 ms in 0 steps since start of marking, biggest step 0.0 ms, walltime since start of marking 899 ms) last resort GC in old space requested
Oct 19 10:53:26 webrtc0-stage node[1408]: [1408:0x33e1bc0]  1354263 ms: Mark-sweep 1405.2 (1436.9) -> 1405.2 (1436.9) MB, 880.2 / 0.0 ms  last resort GC in old space requested
```
